codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java

```diff
@@ -164,6 +166,11 @@ public void readFrame(ChannelHandlerContext ctx, ByteBuf input, Http2FrameListen
             } while (input.isReadable());
         } catch (Http2Exception e) {
             readError = !Http2Exception.isStreamError(e);
+            if (resetReader) {
+                resetReader = false;
+                readingHeaders = true;
+                input.skipBytes(input.readableBytes());
```

Comment:
I think you're right, if there's more data already buffered it'll skip all of it. I'm trying to think of an alternative but if the length of the payload doesn't match the actual payload not sure how we can skip the correct number of bytes to read the next frame correctly.

Reply:
Instead of using a boolean flag we could instead do something like passing `input.readSlice(payloadLength)` into the `processPayloadState` method. That would require us to move the 
```
if (input.readableBytes() < payloadLength) {
    // Wait until the entire payload has been read.
     return;
}
```
check up out of the `processPayloadState` method to right before we slice and call it but tbh that feels more readable anyway.

Doing that has two nice properties and one negative: it's nice because it 'slices' exactly one frame from `input` which (1) makes it hard for `processPayloadState` to corrupt other frames and (2) makes consumption of the frame from `input` happen eagerly so you don't have to worry about skipping bytes on exception. Downside is that it does cost an extra allocation. On balance, I'd happily pay the allocation of the slice wrapper: it isn't a drop in the ocean imo.

Reply:
That makes sense to me, but I think we'd hit some issues if the advertised `payloadLength` of the frame doesn't match the actual payload. In the use case I'm looking at, I see two cases where we could see issues
- __Advertised Payload Length greater than payload__: More data skipped than actually included and next frame would not be properly read.
- __Advertised Payload Length less than payload__: Less data skipped than actually included and data will be left over and next frame would not be properly read.

Reply:
It's convenient to think of the H2 protocol as having (at least) two layers: the framing layer (the first 9 octets of a frame) and then the actual frame payloads themselves which will have some consistency requirements. What I'm suggesting leverages the fact that the length from the framing layer should be authoritative in terms of frame boundaries.
What to do if we detect corruption/inconsistencies in the payload (which could be that the payload is short or long, for example) will be case specific and gets into the details of what @idelpivnitskiy pointed to in his suggestions. This is actually a third advantage of what I'm suggestion: extra bytes can signal corruption and if we slice we can detect that. For example, a RST_STREAM frame with size other than 4 octets should be a connection error (end of [6.4](https://datatracker.ietf.org/doc/html/rfc9113#section-6.4)). However in main if we received a RST_STREAM frame of size 5 (as determined by the frame header) where the last byte is meaningless instead of sending a connection error and potentially alerting the peer to a bug in their implementation we'll instead end up with a corrupt framing layer: we won't fail because we don't enforce the size and we won't consume that extra byte from the data stream so it will end up being interpreted as the first byte of the next frames header.

Reply:
Yeah that's exactly the issue I was seeing, leftover data is parsed as part of the next frame when malformed frames like that are parsed. So with @idelpivnitskiy comments and this thread I'm thinking we can implement the slice mechanism but there needs to be more checks implemented so that the payload length matches the actual payload and for those cases where we don't find a match we can alert by sending a Go Away frame?

Reply:
That sounds right: whether to send a GOAWAY or a RST_STREAM is frame type specific and described (usually at the ends) of the frame definitions. I don't think you have to implement all those checks as part of this PR: they could be done as a follow up. Finding this and making sure we keep a correct frame layer is already a huge step forward from where we are now.

Reply:
Taking a closer look it seems most of them (if not all) already verify the frame length and do the right thing wrt RST vs GOAWAY but its done as part of the `processHeaderState` state so we haven't been able to slice before the exception is thrown, and we may not even have that many bytes available, so we'd need to move the verification to a new method and perform it right before we `processPayloadState(ctx, input.slice(..), ..)`.